File: apps/dsa-web/src/api/workspace.ts
import apiClient from "./index";
import type { AgentCapabilityBindings } from "../types/capabilities";

export type WorkspaceSkill = { id: string; name: string; description: string; category: string; instructions: string; version: number; builtIn: boolean; enabled: boolean };
export type WorkspaceTool = { id: string; name: string; description: string; category: string; enabled: boolean; policy?: Record<string, unknown> };
export type WorkspaceMcpServer = { id: string; name: string; transport: "http" | "stdio"; location: string; credentialKey: string; enabled: boolean; selectable: boolean; healthStatus: string; lastCheckedAt?: string | null; lastError?: string | null; capabilities: Array<{type:string;name:string;description?:string;inputSchema?:Record<string,unknown>}> };
export type WorkspaceDataSourceMember = { id:string; name:string; domain:string; websiteUrl?:string; category:"publisher"|"corporate_wire"|"regulator"; markets:string[] };
export type WorkspaceDataSourceHealth = "not_tested"|"not_configured"|"available"|"degraded"|"unavailable";
export type WorkspaceDataSourceAccessMode = "automatic"|"local"|"no_credential"|"api_key"|"token"|"base_url"|"account"|"custom";
export type WorkspaceDataSource = { id?:number; sourceId:string; name:string; kind:"kline"|"news"|"fundamentals"|"macro"|"other"; description?:string|null; connectionKey:string; required:boolean; builtIn:boolean; selectable:boolean; availability:"system_managed"|"configured"|"unconfigured"|"registered"; selectionMode?:"automatic"|"provider"|"local"; providerName?:string; includedSources?:WorkspaceDataSourceMember[]; markets?:string[]; setupUrl?:string|null; accessMode?:WorkspaceDataSourceAccessMode; configurationKeys?:string[]; healthStatus?:WorkspaceDataSourceHealth; probeSupported?:boolean; operational?:boolean; lastCheckedAt?:string|null; lastLatencyMs?:number|null; lastRecordCount?:number|null; lastErrorCode?:string|null; lastError?:string|null; healthDetail?:Record<string,unknown>; createdAt?:string; updatedAt?:string };
export type WorkspaceExpert = { id: number; avatar?: string | null; name: string; style: string; description: string; philosophy: string; focus: string[]; prompt: string; defaultPrompt: string; version: number; builtIn: boolean; enabled: boolean };
export type WorkspaceExpertTeam = { id: number; name: string; description: string; memberIds: number[]; protocol: string; version: number; builtIn: boolean; enabled: boolean };
export type WorkspaceCapabilityCatalog = { discussionProtocols?: string[]; discussionModes?: string[]; discussionChatBinding?: boolean; discussionReconfiguration?: boolean; skills: WorkspaceSkill[]; tools: WorkspaceTool[]; mcpServers: WorkspaceMcpServer[]; dataSources: WorkspaceDataSource[]; experts: WorkspaceExpert[]; expertTeams: WorkspaceExpertTeam[]; defaults: Record<"chat"|WorkspaceTaskKind, AgentCapabilityBindings> };
export type WorkspaceTaskKind = "research" | "screening" | "trading" | "expert_review" | "market_analysis" | "industry_analysis";
export type WorkspaceTask = { id:string; kind:WorkspaceTaskKind; name:string; market:"CN"|"HK"|"US"|"GLOBAL"; objective:string; subject:Record<string,unknown>; config:Record<string,unknown>; capabilities:AgentCapabilityBindings; version:number; enabled:boolean; createdAt:string; updatedAt:string };
export type WorkspaceArtifact = { id:string; type:string; title:string; content:unknown; text?:string|null; version:number; createdAt:string };
export type WorkspaceRun = { primaryReportRunId?:string; relatedRunIds?:string[]; reportTitle?:string; outcome?:{status:"pending"|"produced"|"empty"|"proposal"|"partial"|"blocked"|"unverified"|"failed"|"cancelled";message:string}; id:string; taskId:string; kind:WorkspaceTaskKind; status:"queued"|"running"|"completed"|"failed"|"cancelled"; triggerType:string; dataSnapshotId:string; taskSnapshot:WorkspaceTask; resultSummary?:Record<string,unknown>|null; errorCode?:string|null; errorMessage?:string|null; startedAt?:string|null; completedAt?:string|null; createdAt:string; updatedAt:string; artifacts:WorkspaceArtifact[]; dataSnapshot?:{id:string;asOf:string;sourceIds:string[];sourceVersions:Record<string,unknown>;quality:Record<string,unknown>}|null };
export type WorkspaceSchedule = { id:string; taskId:string; name:string; scheduleMode:"daily"|"interval"; runAt?:string|null; intervalDays?:number; intervalMinutes?:number|null; timezone:string; enabled:boolean; nextRunAt:string; lastRunAt?:string|null; lastRunId?:string|null; createdAt:string; updatedAt:string };
export type MarketDashboardWidgetId = "overview"|"macro"|"indices"|"breadth"|"sectors"|"news"|"subscriptions";
export type MarketArtifactSummary = { text:string; risks:string[]; confidence?:number|null };
export type MarketSubscription = {
  id:string;
  taskId:string;
  market:WorkspaceTask["market"];
  title:string;
  enabled:boolean;
  position:number;
  task:WorkspaceTask|null;
  latestRun:{id:string;status:WorkspaceRun["status"];errorCode?:string|null;errorMessage?:string|null;startedAt?:string|null;completedAt?:string|null;createdAt:string}|null;
  latestArtifact:{id:string;runId:string;type:string;title:string;summary:MarketArtifactSummary;createdAt:string;dataSnapshotId?:string|null}|null;
  schedules:WorkspaceSchedule[];
  createdAt:string;
  updatedAt:string;
};
export type MarketDashboard = { market:WorkspaceTask["market"];widgetIds:MarketDashboardWidgetId[];newsSourceIds:number[];newsKeywords:string[];subscriptions:MarketSubscription[];updatedAt?:string|null };

const root = "/api/v1/workspace";

export type DefaultTaskPlan = {
  policyVersion: string;
  task: Pick<WorkspaceTask, "kind" | "name" | "market" | "objective" | "subject" | "config" | "capabilities">;
  strategyName: string;
  skillNames: string[];
  teamName: string;
  expertCount: number;
  reasons: string[];
  warnings: string[];
  notice: string;
};

export const workspaceApi = {
  async getDefaultTaskPlan(kind: "research" | "screening" | "trading", market = "CN", stock?: string) { const { data } = await apiClient.get<DefaultTaskPlan>(`${root}/default-task-plan`, { params: { kind, market, stock } }); return data; },
  async getCapabilities() { const { data } = await apiClient.get<WorkspaceCapabilityCatalog>(`${root}/capabilities`); return data; },
  async listSkills() { const { data } = await apiClient.get<WorkspaceSkill[]>(`${root}/skills`); return data; },
  async createSkill(payload: Omit<WorkspaceSkill,"version"|"builtIn">) { const { data } = await apiClient.post<WorkspaceSkill>(`${root}/skills`, payload); return data; },
  async updateSkill(id:string,payload:Partial<WorkspaceSkill>) { const { data } = await apiClient.patch<WorkspaceSkill>(`${root}/skills/${id}`,payload); return data; },
  async deleteSkill(id:string) { await apiClient.delete(`${root}/skills/${id}`); },
  async listTools() { const { data } = await apiClient.get<WorkspaceTool[]>(`${root}/tools`); return data; },
  async listDataSources() { const { data } = await apiClient.get<WorkspaceDataSource[]>(`${root}/data-sources`); return data; },
  async probeDataSource(id:string) { const { data } = await apiClient.post<WorkspaceDataSource>(`${root}/data-sources/${encodeURIComponent(id)}/probe`); return data; },
  async createDataSource(payload:{name:string;description?:string;connectionKey:string;setupUrl?:string;accessMode:Exclude<WorkspaceDataSourceAccessMode,"automatic"|"local">;kind:WorkspaceDataSource["kind"];markets:string[]}) { const { data } = await apiClient.post<WorkspaceDataSource>(`${root}/data-sources`,payload); return data; },
  async archiveDataSource(id:number) { const { data } = await apiClient.delete<WorkspaceDataSource & {archived:boolean}>(`${root}/data-sources/${id}`); return data; },
  async setPreferences(kind:string,enabledIds:Array<string|number>) { const { data } = await apiClient.put(`${root}/capabilities/${kind}/preferences`,{enabledIds}); return data; },
  async listMcpServers() { const { data } = await apiClient.get<WorkspaceMcpServer[]>(`${root}/mcp-servers`); return data; },
  async createMcpServer(payload:{name:string;transport:"http"|"stdio";location:string;credentialKey:string;enabled?:boolean}) { const { data } = await apiClient.post<WorkspaceMcpServer>(`${root}/mcp-servers`,payload); return data; },
  async updateMcpServer(id:string,payload:Partial<WorkspaceMcpServer>) { const { data } = await apiClient.patch<WorkspaceMcpServer>(`${root}/mcp-servers/${id}`,payload); return data; },
  async deleteMcpServer(id:string) { await apiClient.delete(`${root}/mcp-servers/${id}`); },
  async probeMcpServer(id:string) { const { data } = await apiClient.post<WorkspaceMcpServer>(`${root}/mcp-servers/${id}/probe`); return data; },
  async listExperts() { const { data } = await apiClient.get<WorkspaceExpert[]>(`${root}/experts`); return data; },
  async createExpert(payload:{name:string;avatar?:string|null;style:string;description?:string;philosophy?:string;focus?:string[];prompt:string}) { const { data } = await apiClient.post<WorkspaceExpert>(`${root}/experts`,payload); return data; },
  async updateExpert(id:number,payload:Partial<WorkspaceExpert>) { const { data } = await apiClient.patch<WorkspaceExpert>(`${root}/experts/${id}`,payload); return data; },
  async deleteExpert(id:number) { await apiClient.delete(`${root}/experts/${id}`); },
  async listExpertTeams() { const { data } = await apiClient.get<WorkspaceExpertTeam[]>(`${root}/expert-teams`); return data; },
  async createExpertTeam(payload:{name:string;description?:string;memberIds:number[];protocol:string}) { const { data } = await apiClient.post<WorkspaceExpertTeam>(`${root}/expert-teams`,payload); return data; },
  async updateExpertTeam(id:number,payload:Partial<WorkspaceExpertTeam>) { const { data } = await apiClient.patch<WorkspaceExpertTeam>(`${root}/expert-teams/${id}`,payload); return data; },
  async deleteExpertTeam(id:number) { await apiClient.delete(`${root}/expert-teams/${id}`); },
  async createTask(payload:{kind:WorkspaceTaskKind;name:string;market:WorkspaceTask["market"];objective:string;subject?:Record<string,unknown>;config?:Record<string,unknown>;capabilities:AgentCapabilityBindings;enabled?:boolean}) { const { data } = await apiClient.post<WorkspaceTask>(`${root}/tasks`,payload); return data; },
  async updateTask(id:string,payload:Partial<WorkspaceTask>) { const { data } = await apiClient.patch<WorkspaceTask>(`${root}/tasks/${id}`,payload); return data; },
  async listTasks(kind?:WorkspaceTaskKind) { const { data } = await apiClient.get<WorkspaceTask[]>(`${root}/tasks`,{params:{kind}}); return data; },
  async deleteTask(id:string) { await apiClient.delete(`${root}/tasks/${id}`); },
  async runTask(id:string) { const { data } = await apiClient.post<WorkspaceRun>(`${root}/tasks/${id}/runs`,{triggerType:"manual"}); return data; },
  async listRuns(kind?:WorkspaceTaskKind) { const { data } = await apiClient.get<WorkspaceRun[]>(`${root}/runs`,{params:{kind}}); return data; },
  async getRun(id:string) { const { data } = await apiClient.get<WorkspaceRun>(`${root}/runs/${id}`); return data; },
  async cancelRun(id:string) { const { data } = await apiClient.post(`${root}/runs/${id}/cancel`); return data; },
  async createSchedule(payload:{taskId:string;name:string;scheduleMode:"daily"|"interval";runAt?:string;intervalDays?:number;intervalMinutes?:number;timezone:string;enabled?:boolean;publishToMarket?:boolean;marketDashboardTitle?:string}) { const { data } = await apiClient.post<WorkspaceSchedule>(`${root}/schedules`,payload); return data; },
  async listSchedules() { const { data } = await apiClient.get<WorkspaceSchedule[]>(`${root}/schedules`); return data; },
  async updateSchedule(id:string,payload:Partial<WorkspaceSchedule>) { const { data } = await apiClient.patch<WorkspaceSchedule>(`${root}/schedules/${id}`,payload); return data; },
  async deleteSchedule(id:string) { await apiClient.delete(`${root}/schedules/${id}`); },
  async getMarketDashboard(market:WorkspaceTask["market"]) { const { data } = await apiClient.get<MarketDashboard>(`${root}/market-dashboards/${market}`); return data; },
  async updateMarketDashboard(market:WorkspaceTask["market"],payload:Pick<MarketDashboard,"widgetIds"|"newsSourceIds"|"newsKeywords">) { const { data } = await apiClient.put<MarketDashboard>(`${root}/market-dashboards/${market}`,payload); return data; },
  async listMarketSubscriptions(market?:WorkspaceTask["market"]) { const { data } = await apiClient.get<MarketSubscription[]>(`${root}/market-subscriptions`,{params:{market}}); return data; },
  async createMarketSubscription(payload:{taskId:string;market:WorkspaceTask["market"];title?:string;enabled?:boolean}) { const { data } = await apiClient.post<MarketSubscription>(`${root}/market-subscriptions`,payload); return data; },
  async updateMarketSubscription(id:string,payload:{title?:string;enabled?:boolean;position?:number}) { const { data } = await apiClient.patch<MarketSubscription>(`${root}/market-subscriptions/${id}`,payload); return data; },
  async deleteMarketSubscription(id:string) { await apiClient.delete(`${root}/market-subscriptions/${id}`); },
};

File: api/v1/schemas/workspace.py
# -*- coding: utf-8 -*-
"""API contracts for the Agent-first financial workspace."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator
from src.services.expert_avatar import validate_avatar


class WorkspaceModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class CapabilityBindings(WorkspaceModel):
    skillIds: List[str] = Field(default_factory=list, max_length=50)
    toolIds: List[str] = Field(default_factory=list, max_length=100)
    mcpIds: List[str] = Field(default_factory=list, max_length=30)
    dataSourceIds: List[str] = Field(default_factory=list, max_length=30)
    expertIds: List[int] = Field(default_factory=list, max_length=20)
    expertTeamIds: List[int] = Field(default_factory=list, max_length=10)


class CapabilityPreferenceRequest(WorkspaceModel):
    enabledIds: List[str | int] = Field(default_factory=list, max_length=200)


class PortfolioResearchRules(WorkspaceModel):
    lossPct: float = Field(10, ge=0.1, le=100, allow_inf_nan=False)
    profitPct: float = Field(20, ge=0.1, le=100, allow_inf_nan=False)
    dailyMovePct: float = Field(5, ge=0.1, le=100, allow_inf_nan=False)


class PortfolioResearchRequest(WorkspaceModel):
    strategyVersionId: Optional[int] = Field(None, gt=0)
    capabilities: Optional[CapabilityBindings] = None
    rules: Optional[PortfolioResearchRules] = None
    dailyEnabled: bool = False
    intervalDays: Optional[int] = Field(None, strict=True, ge=1, le=365)
    runAt: Optional[str] = Field(None, pattern=r"^(?:[01]\d|2[0-3]):[0-5]\d$")


class DataSourceCreateRequest(WorkspaceModel):
    name: str = Field(..., min_length=1, max_length=120)
    description: Optional[str] = Field(None, max_length=1000)
    connectionKey: str = Field(..., min_length=2, max_length=160)
    setupUrl: Optional[str] = Field(None, max_length=2048)
    accessMode: Literal["no_credential", "api_key", "token", "base_url", "account", "custom"] = "custom"
    kind: Literal["kline", "news", "fundamentals", "macro", "other"]
    markets: List[Literal["cn", "hk", "us"]] = Field(..., min_length=1, max_length=3)


class SkillCreateRequest(WorkspaceModel):
    id: Optional[str] = Field(None, min_length=2, max_length=128, pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]+$")
    name: str = Field(..., min_length=1, max_length=120)
    category: Literal["research", "screening", "risk", "trading", "general"] = "general"
    description: str = Field("", max_length=4000)
    instructions: str = Field(..., min_length=1, max_length=30000)
    enabled: bool = True


class SkillUpdateRequest(WorkspaceModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    category: Optional[Literal["research", "screening", "risk", "trading", "general"]] = None
    description: Optional[str] = Field(None, max_length=4000)
    instructions: Optional[str] = Field(None, min_length=1, max_length=30000)
    enabled: Optional[bool] = None


class McpServerCreateRequest(WorkspaceModel):
    id: Optional[str] = Field(None, min_length=2, max_length=128, pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]+$")
    name: str = Field(..., min_length=1, max_length=120)
    transport: Literal["http", "stdio"] = "http"
    location: str = Field(..., min_length=1, max_length=4000)
    credentialKey: str = Field("", max_length=160)
    enabled: bool = True


class McpServerUpdateRequest(WorkspaceModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    transport: Optional[Literal["http", "stdio"]] = None
    location: Optional[str] = Field(None, min_length=1, max_length=4000)
    credentialKey: Optional[str] = Field(None, max_length=160)
    enabled: Optional[bool] = None


class ExpertCreateRequest(WorkspaceModel):
    avatar: Optional[str] = Field(None, max_length=180000)
    _validate_avatar = field_validator("avatar")(validate_avatar)
    name: str = Field(..., min_length=1, max_length=120)
    style: str = Field("自定义投资视角", min_length=1, max_length=240)
    description: str = Field("", max_length=4000)
    philosophy: str = Field("", max_length=4000)
    focus: List[str] = Field(default_factory=list, max_length=30)
    prompt: str = Field(..., min_length=1, max_length=30000)
    enabled: bool = True


class ExpertUpdateRequest(WorkspaceModel):
    avatar: Optional[str] = Field(None, max_length=180000)
    _validate_avatar = field_validator("avatar")(validate_avatar)
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    style: Optional[str] = Field(None, min_length=1, max_length=240)
    description: Optional[str] = Field(None, max_length=4000)
    philosophy: Optional[str] = Field(None, max_length=4000)
    focus: Optional[List[str]] = Field(None, max_length=30)
    prompt: Optional[str] = Field(None, min_length=1, max_length=30000)
    enabled: Optional[bool] = None


class ExpertTeamCreateRequest(WorkspaceModel):
    name: str = Field(..., min_length=1, max_length=120)
    description: str = Field("", max_length=4000)
    memberIds: List[int] = Field(..., min_length=1, max_length=20)
    protocol: str = Field(..., min_length=1, max_length=10000)
    enabled: bool = True


class ExpertTeamUpdateRequest(WorkspaceModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    description: Optional[str] = Field(None, max_length=4000)
    memberIds: Optional[List[int]] = Field(None, min_length=1, max_length=20)
    protocol: Optional[str] = Field(None, min_length=1, max_length=10000)
    enabled: Optional[bool] = None


class TaskCreateRequest(WorkspaceModel):
    kind: Literal["research", "screening", "trading", "expert_review", "market_analysis", "industry_analysis"]
    name: str = Field(..., min_length=1, max_length=160)
    market: Literal["CN", "HK", "US", "GLOBAL"] = "CN"
    objective: str = Field(..., min_length=1, max_length=30000)
    subject: Dict[str, Any] = Field(default_factory=dict)
    config: Dict[str, Any] = Field(default_factory=dict)
    capabilities: CapabilityBindings = Field(default_factory=CapabilityBindings)
    enabled: bool = True


class TaskUpdateRequest(WorkspaceModel):
    name: Optional[str] = Field(None, min_length=1, max_length=160)
    market: Optional[Literal["CN", "HK", "US", "GLOBAL"]] = None
    objective: Optional[str] = Field(None, min_length=1, max_length=30000)
    subject: Optional[Dict[str, Any]] = None
    config: Optional[Dict[str, Any]] = None
    capabilities: Optional[CapabilityBindings] = None
    enabled: Optional[bool] = None


class RunCreateRequest(WorkspaceModel):
    triggerType: Literal["manual", "schedule", "retry"] = "manual"


class ScheduleCreateRequest(WorkspaceModel):
    intervalDays: int = Field(1, strict=True, ge=1, le=365)
    taskId: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=160)
    scheduleMode: Literal["daily", "interval"]
    runAt: Optional[str] = Field(None, max_length=8)
    intervalMinutes: Optional[int] = Field(None, ge=5, le=10080)
    timezone: str = Field("Asia/Shanghai", min_length=1, max_length=64)
    enabled: bool = True
    publishToMarket: bool = False
    marketDashboardTitle: Optional[str] = Field(None, min_length=1, max_length=160)


class ScheduleUpdateRequest(WorkspaceModel):
    intervalDays: Optional[int] = Field(None, strict=True, ge=1, le=365)
    name: Optional[str] = Field(None, min_length=1, max_length=160)
    runAt: Optional[str] = Field(None, max_length=8)
    intervalMinutes: Optional[int] = Field(None, ge=5, le=10080)
    timezone: Optional[str] = Field(None, min_length=1, max_length=64)
    enabled: Optional[bool] = None


class MarketDashboardUpdateRequest(WorkspaceModel):
    widgetIds: List[Literal["overview", "macro", "indices", "breadth", "sectors", "news", "subscriptions"]] = Field(..., min_length=1, max_length=7)
    newsSourceIds: List[int] = Field(default_factory=list, max_length=100)
    newsKeywords: List[str] = Field(default_factory=list, max_length=20)


class MarketSubscriptionCreateRequest(WorkspaceModel):
    taskId: str = Field(..., min_length=1, max_length=64)
    market: Literal["CN", "HK", "US", "GLOBAL"]
    title: Optional[str] = Field(None, min_length=1, max_length=160)
    enabled: bool = True


class MarketSubscriptionUpdateRequest(WorkspaceModel):
    title: Optional[str] = Field(None, min_length=1, max_length=160)
    enabled: Optional[bool] = None
    position: Optional[int] = Field(None, ge=0, le=10000)
